### PR TITLE
Add dynamic README header image endpoint

### PR DESCRIPTION
**Pull request by AlemTuzlak**
Library repo READMEs each open with a hand-made banner PNG committed to `media/header_*.png`. They're inconsistent in size (query is 3000×1704, table is 1728×874), too tall for a README, and updating the logo or a tagline means re-exporting images across a dozen repos.

This adds `GET /api/readme/<libraryId>.png`, which renders that banner on demand at **1800×450** in **light and dark**, reusing the takumi renderer, brand assets and per-category accent colors that already back the OG cards. Repos point a `<picture>` at the URL instead of committing a file — so a branding change lands in every README at once, in both themes.

## What it looks like

Shown at 900px, the width GitHub renders README images at. Light then dark for each.

### TanStack Query

TanStack Query light
TanStack Query dark

### TanStack Router

TanStack Router light
TanStack Router dark

### TanStack Pacer

TanStack Pacer light
TanStack Pacer dark

Per-package READMEs add `?framework=`, so `packages/react-start/README.md` gets its own name:

TanStack React Start
TanStack Solid Start

<details>
<summary>Every other library, both themes</summary>

**ai**

ai light
ai dark

**charts**

charts light
charts dark

**cli**

cli light
cli dark

**config**

config light
config dark

**db**

db light
db dark

**devtools**

devtools light
devtools dark

**form**

form light
form dark

**highlight**

highlight light
highlight dark

**hotkeys**

hotkeys light
hotkeys dark

**intent**

intent light
intent dark

**markdown**

markdown light
markdown dark

**mcp**

mcp light
mcp dark

**ranger**

ranger light
ranger dark

**start**

start light
start dark

**store**

store light
store dark

**table**

table light
table dark

**virtual**

virtual light
virtual dark

**workflow**

workflow light
workflow dark

</details>

## Usage

```html
<!-- TanStack/query → README.md -->
<picture>
  <source media="(prefers-color-scheme: dark)"
          srcset="https://tanstack.com/api/readme/query.png?theme=dark" />
  <source media="(prefers-color-scheme: light)"
          srcset="https://tanstack.com/api/readme/query.png" />
  <img src="https://tanstack.com/api/readme/query.png" alt="TanStack Query" width="900" />
</picture>
```

Per [GitHub's own guidance](https://github.blog/developer-skills/github/how-to-make-your-images-in-markdown-on-github-adjust-for-dark-mode-and-light-mode/). The trailing `` stays the light variant as the fallback for renderers that ignore `<picture>` (npm, most editors). The post's `#gh-dark-mode-only` fragment trick is documented but not recommended — it renders both images in any client that doesn't special-case it.

## Parameters

| Param | Behavior |
| --- | --- |
| path splat | Library id. Unknown → `404` |
| `?framework=` | Must be one of the library's frameworks. Anything else → `400` listing the accepted values |
| `?theme=` | `light` (default) or `dark`. Anything else → `400` |
| `?title=` | Replaces the name (80 char clamp). Wins over `?framework=` |
| `?subtitle=` | Replaces the tagline (160 char clamp) |

Supplied-but-invalid is always a 400 rather than a silent fallback, so a typo in a README surfaces as a broken image during review instead of a banner that quietly says the wrong thing.

## Dark palette

Taken from the same `app.css` token sets the site uses, so a dark banner matches the site's own dark mode: `background-default` for the surface, `text-secondary` for the prefix and tagline, and the category **300** step for the accent.

One deliberate deviation, visible in the devtools banner above: the site's dark tooling accent is `--color-ds-neutral-200`, which is *also* the dark `text-secondary` value — so devtools/workflow/cli/mcp/config would render name and tagline in a single colour. The dark tooling accent uses the neutral **tint** step instead.

Worth flagging that **light mode has the same collision today** — its tooling accent and secondary text are both `#3e3529`, so those banners are flat in light mode (compare the two devtools images above). Its palette is pinned by an existing test and shared with the live 1200×630 social cards, so I left it alone rather than changing it as a side effect here. Happy to fix it in a follow-up if you want the light tooling banners to have hierarchy too.

## Notes for reviewers

- The render path (asset loading, fonts, takumi module) is now **shared** between the OG cards and the README headers rather than duplicated — `generateOgImageResponse` is unchanged from the outside, and `getAccentColor` keeps its old signature via a defaulted `theme` argument.
- Emblem rasters are now transparent rather than cream-backed, so one file serves either surface; a cream emblem is generated alongside the charcoal one.
- Text is bounded to the canvas: `maxWidth` handles wrapping, and since takumi supports neither `wordBreak` nor `overflowWrap`, a single unbreakable token is scaled to fit. The fitting is a character-count estimate with deliberately pessimistic ratios — takumi exposes no text-measurement API — so a bad guess can only shrink a line, never overflow.
- `pnpm run readme:preview` renders every library in both themes plus every framework variant and a set of clamp-length boundary cases. The preview PNGs above live on the `readme-header-previews` branch, deliberately off this one so the diff carries no generated images.
- Docs: `docs/readme-headers.md`.
- `tests/og-branding.test.ts` covers the dark palette, light/dark parity, theme validation and the accent/text collision rule.
- One commit is unrelated to the feature: `tests/local-repo-path.test.ts` hardcoded POSIX paths and failed on Windows, blocking the pre-commit hook. Made it platform-agnostic.

**~25% dead space on the right** is inherent to a left-aligned 4:1 banner. Happy to drop to 1600×450 or center the group if reviewers prefer.

## Summary by CodeRabbit

* **New Features**
  * Added README header OG image API with optional light/dark theme support and theme-aware rendering, plus improved preview gallery.
  * Generated branded emblem assets for both charcoal and cream variants.
* **Documentation**
  * Updated README header image guidance for light/dark usage and the local preview workflow.
* **Bug Fixes**
  * Added validation for unknown/blank theme values and improved caption wrapping for long text.
* **Tests**
  * Expanded theme/dark-mode OG branding coverage and improved local test path portability.

### CODE REVIEW

File: src/routes/api/readme/{$}[.]png.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        if (!library) {
+          return new Response(`Unknown library: ${libraryId}`, { status: 404 })
+        }
+
+        const url = new URL(request.url)
+
+        // Validated rather than ignored: a typo'd framework in a README should
+        // surface as a broken image, not as a banner naming the wrong package.
+        const framework = url.searchParams.get('framework') ?? undefined
+        if (framework && !library.frameworks.includes(framework as Framework)) {
```
[the remark this answers is not in the document]
Comment (AlemTuzlak) [reply src/routes/api/readme/{$}[.]png.ts]: Confirmed and fixed in 33ea3370.

`?framework=` produced `''`, which is falsy, so it skipped the guard and returned a default banner with 200 — the endpoint documents a 400 for any supplied-but-unsupported value. Switched to `framework !== undefined`, so an omitted parameter still means "no framework" while a blank one is treated as the malformed URL it is.

Verified against a dev server:

| Request | Before | After |
| --- | --- | --- |
| `?framework=` | 200 (default banner) | 400 `Unknown framework "" for TanStack Start. Expected one of: react, solid` |
| `?framework=react` | 200 | 200 |
| no `framework` param | 200 | 200 |
| `?framework=vue` | 400 | 400 |

File: src/server/og/readme-template.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+          }}
+        >
+          {nameLine2 ? <span style={{ fontSize: 44 }}>{nameLine1}</span> : null}
+          <span style={{ fontSize: 96, color: props.accentColor }}>
+            {nameLine2 || nameLine1}
+          </span>
+        </div>
+        <div style={{ fontSize: 30, fontWeight: 400, lineHeight: 1.3 }}>
+          {props.tagline}
+        </div>
```
[the remark this answers is not in the document]
Comment (AlemTuzlak) [reply src/server/og/readme-template.tsx:164]: Confirmed and fixed in 33ea3370 — the 1200x630 template already caps its copy with `maxWidth`, and this one had no bound at all.

Two parts, because `maxWidth` alone wasn't enough:

1. `maxWidth: TEXT_MAX_WIDTH` (the space actually left beside the emblem, derived from the layout constants) on the text column, name block and tagline. That fixes text with spaces — it now wraps.
2. takumi supports neither `wordBreak` nor `overflowWrap` (grepped `@takumi-rs/core` and `/helpers` — no such style keys), so a single long token has nowhere to break and still ran to the edge. Those lines now scale down to fit.

The fitting applies two independent constraints, since a lone token can't use a second line even when the tagline is allowed to wrap:

```
fitted = min(baseSize,
             budget(lines) / (text.length   * ratio),   // whole string, wrapped
             budget(1)     / (longestToken  * ratio))   // any one token, alone
```

It's a character-count estimate, not real metrics — takumi exposes no measurement API — so the ratios are deliberately pessimistic (near `W`/`M` width rather than average). A bad guess can only shrink a line, never overflow. Noted in a comment with the upgrade path.

Boundary renders added to `readme-header-preview.ts` per your suggestion — clamp-length spaced text, the same length as one unbreakable token, and the widest glyphs in the set:

**80/160 chars, spaced** — wraps, was overflowing before:

spaced

**80/160 chars, single token:**

single token

**80/160 chars of `W`** — the case most likely to defeat the estimate:

widest glyphs

All three stay inside the canvas. Re-rendered every real library to confirm no regression — none trigger the fitting, so they're byte-identical in layout to before.

File: scripts/readme-header-preview.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      await renderToFile(file, {
+        libraryId: 'query',
+        title: boundary.title,
+        subtitle: boundary.subtitle,
+      })
+    ) {
+      boundaryFiles.push({
+        file,
+        url: `/api/readme/query.png?title=${encodeURIComponent(boundary.title).slice(0, 40)}…&subtitle=…`,
+      })
```
[the remark this answers is not in the document]
Comment (AlemTuzlak) [reply scripts/readme-header-preview.ts]: Fair — fixed in 8fd4c4fb. The caption now carries the full encoded query string built with `URLSearchParams`, so it can be pasted verbatim to regenerate the exact image. Added `overflow-wrap: anywhere` to the gallery's `figcaption` so the long ones wrap instead of stretching the page.

File: src/server/og/readme-template.tsx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  const limit = (chars: number, budget: number) =>
+    chars === 0 ? baseSize : budget / (chars * avgGlyphRatio)
+
+  const fitted = Math.min(
+    baseSize,
+    limit(text.length, TEXT_MAX_WIDTH * lines),
+    limit(longestToken, TEXT_MAX_WIDTH),
+  )
+
+  return Math.max(16, Math.floor(fitted))
```
[the remark this answers is not in the document]
Comment (AlemTuzlak) [reply src/server/og/readme-template.tsx]: Good catch — the floor was above the range the calculation actually produces, so for exactly the inputs `fitFontSize` exists to handle, the floor decided the size instead. Fixed in 8fd4c4fb: lowered to 8px, which now only guards against a zero or negative result.

Chasing it further turned up the real problem behind your number. With the floor out of the way the ratios were still optimistic for the widest glyphs — 0.78/0.66 sat above *average* advance, not near the *maximum*. They're now 0.95 (Bricolage Bold) and 0.82 (Inter), so the estimate is pessimistic against an all-`W` input rather than merely conservative.

Where that leaves the worst case, at 900px as GitHub renders it:

widest glyphs

The 80-char title line now clears the padding with real margin. The 160-char all-`W` tagline lands at 10px and still runs ~24px past the padding line, inside the canvas but not inside the padding — I stopped there deliberately rather than shrink further, since text is already illegible at that size and the guarantee worth holding is "never bleeds outside the image". Shrinking to 8px to satisfy a metric nobody can read seemed like the wrong trade; happy to take it to zero tolerance if you disagree.

Worth being explicit that this stays an estimate: takumi exposes no text-measurement API, so the ratios are calibrated by eye against rendered output, and the comment says so along with the upgrade path. Re-rendered all 21 libraries and every framework variant — none reach the fitting path, so real banners are unchanged.
